channels/applicationinsights-channel-js/src/Sender.ts

```diff
@@ -239,18 +244,77 @@ export class Sender extends BaseTelemetryPlugin implements IChannelControlsAI {
+                    // This is so any redirect endpointUrl is not overwritten
+                    if (_orgEndpointUrl !== senderConfig.endpointUrl) {
+                        if (_orgEndpointUrl) {
+                            // TODO: add doc to remind users to flush before changing endpoint
+                            _self._buffer.clear();
+                        }
+                        _endpointUrl = _orgEndpointUrl = senderConfig.endpointUrl;
+                    }
+
+                    if (_customHeaders && _customHeaders !== senderConfig.customHeaders) {
+                        // Removing any previously defined custom headers as they have changed
+                        arrForEach(_customHeaders, customHeader => {
+                            delete _headers[customHeader.header];
+                        });
+                    }
+
                     _maxBatchSizeInBytes = senderConfig.maxBatchSizeInBytes;
                     _beaconSupported = (senderConfig.onunloadDisableBeacon === false || senderConfig.isBeaconApiDisabled === false) && isBeaconsSupported();
-
+                    
+                    let canUseSessionStorage = !!senderConfig.enableSessionStorageBuffer && utlCanUseSessionStorage();
+                    let namePrefix = senderConfig.namePrefix;
+                    let emitLineDelimitedJson = senderConfig.emitLineDelimitedJson;
+                    let eventsLimitInMem = senderConfig.eventsLimitInMem;
+                    let shouldUpdate = (canUseSessionStorage !== _sessionStorageUsed)
+                                    || (canUseSessionStorage && (_namePrefix !== namePrefix)) // prefixName is only used in session storage
+                                    || (_emitLineDelimitedJson !== emitLineDelimitedJson ) //emitLineDelimitedJson is used in buffer.size()
+                                    || (!canUseSessionStorage && (_eventsLimitInMem !== eventsLimitInMem)); // eventsLimitInMem is only used in memory storage
+                    
                     if (_self._buffer) {
-                        // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-                        // TODO 3.x.x !!! - Handle the enableSessionStorageBuffer changing
-                        // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+                        // case1 (Pre and Now enableSessionStorageBuffer settings are same)
+                        // if namePrefix or eventsLimitInMem or emitLineDelimitedJson is changed, transfer current buffer to new buffer
+                        // else no action needed
+
+                        //case2 (Pre and Now enableSessionStorageBuffer settings are changed)
+                        // transfer current buffer to new buffer
+                       
+                        if (shouldUpdate) {
+                            _self.pause();
+                            let unsentPayload = _self._buffer.getItems().slice(0);
```

Comment:
We also need to track the "sent" but not acknowledged items and "move" them as well as if the request that sent them fails they will also need to be re-sent. It might be easier to add a `copy` function to the Buffer implementations that way we can test them also.